### PR TITLE
rgw_sal_motr: [CORTX-31066] Fix multipart object download

### PR DESCRIPTION
**Pull request by DPG17**
Problem:
The download of multipart object fails when the part size is not
multiple of 4MB.
Due to logical issue in code, during object write, the last IO chunk is written to
the already written offset; thereby causing less object size. 
When such object is read, it tries to read more data than available, 
causing Motr to return -ENOENT.

Solution:
Fixed logical issue in code so that last IO write is done in the correct
offset/region of object space.

Signed-off-by: Dattaprasad Govekar <dattaprasad.govekar@seagate.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by andriytk** [comment]
@siningwuseagate, please take a look.

**Comment by DPG17** [comment]
> > Due to logical issue in code, during object write, the last IO chunk is written to
> > the already written offset
> 
> I don't see how it was possible from the old code. Could you please elaborate it?

Not in original code, but the code written in #154

**Comment by bkirunge7** [comment]
@DPG17  can you elaborate solution, what was the logical issue

**Comment by DPG17** [comment]
> @DPG17 can you elaborate solution, what was the logical issue

The issue is during write IO (i.e. Part PUT object), the last IO was overwritten on the existing offset/region in object. E.g. if object is 10MB size, there are total three write IO: (1) 4MB written to offset 0. (2) Next 4MB written to offset 4MB. (3) last 4MB was also written to 4MB. Actually it should have be written at offset 8MB.

File: src/rgw/rgw_sal_motr.cc [hunk 1]
```diff
@@ -1906,9 +1906,13 @@ int MotrObject::write_mobj(const DoutPrefixProvider *dpp, bufferlist&& in_buffer
       io_ctxt.total_bufer_sz += left;
       return 0;
     } else {
-      // Append last buffer
-      io_ctxt.accumulated_buffer_list.push_back(std::move(data));
-      io_ctxt.total_bufer_sz += left;
+      // This is last IO. Check if we have previously accumulated buffers.
```
Comment (bkirunge7) [inline src/rgw/rgw_sal_motr.cc:1909]: when accumulated buffer reaches to size of bs its needs to be written to MOTR immediately before reading next block of data

**Review by sachinpunadikar — APPROVED** [review]
+1

**Review by bkirunge7 — APPROVED** [review]

**Pull request merged**